### ISSUE TITLE
Multi-GPU training and mixed precision

### ISSUE THREAD

**Issue by eliotwalt** [microsoft/aurora#125]
Hi, thanks for this super nice work! 

I'm trying to train Aurora on multiple 95GB H100s but keep running into issues with mixed precision.

## Environment

- Python version: Python 3.11.7
- Pytorch version: 2.5.1+cu124
- Aurora version: 1.7.0

## Summary of the problem

I am unable to train Aurora (large) on more than 1 GPU. I have tried the suggestions in https://github.com/microsoft/aurora/issues/121 and https://github.com/microsoft/aurora/issues/119, namely using `autocast` rather than `bf16-mode` and changing the activation checkpointing but, regardless of the configuration, I eventually get either a Cuda OOM or an illegal memory access error.

I have actually created a repo to test these different configurations on our cluster using random data. It is publicly available [here](https://github.com/eliotwalt/aurora-testbed) and contains a much more complete description of the problem. Being relatively new to distributed training, I might be missing. Please find below a simple training script and the torchrun command I use to start it. 

## Training script

```python
import os
from tqdm import tqdm
import datetime
import contextlib
import argparse

import torch
import torch.nn as nn
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.utils.data.distributed import DistributedSampler
from torch.utils.data import DataLoader, TensorDataset
from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
    apply_activation_checkpointing,
)

from aurora.model.aurora import Aurora, AuroraSmallPretrained
from aurora import Batch, Metadata

torch.backends.cuda.matmul.allow_tf32 = True
torch.set_float32_matmul_precision("high")

# https://github.com/microsoft/aurora/issues/80#issuecomment-2849357112
os.environ["PYTORCH_CUDA_ALLOC_CONF"]="max_split_size_mb:32"

def make_batch(x, lat, lon):
    H, W = x.shape[-2:]
    return Batch(
        surf_vars={var: x[:,:,i] for i, var in enumerate(["10u", "10v", "2t", "msl"])},
        atmos_vars={var: x[:,:,i*13:(i+1)*13] for i, var in enumerate(["t", "u", "v", "q", "z"])},
        static_vars={var: torch.randn(H,W) for var in ["z", "lsm", "slt"]},
        metadata=Metadata(
            lat=lat, lon=lon, time=(datetime.datetime.now(),),
            atmos_levels=(50, 100, 150, 200, 250, 300, 400, 500, 600, 700, 850, 925, 1000),
        )
    )
    
def compute_loss(criterion, preds, target):
    loss = None
    for var_kind in ["surf_vars", "atmos_vars"]:
        for var_name in getattr(target, var_kind).keys():
            pred_var = getattr(preds, var_kind)[var_name]
            target_var = getattr(target, var_kind)[var_name]
            if loss is None: loss = criterion(pred_var, target_var)
            else: loss += criterion(pred_var, target_var)
    return loss

def custom_activation_checkpointing(model, module_names=("Perceiver3DEncoder","Swin3DTransformerBackbone","Basic3DEncoderLayer","Basic3DDecoderLayer","Perceiver3DDecoder","LinearPatchReconstruction")):
        def check(x: torch.nn.Module) -> bool:
            name = x.__class__.__name__
            return name in module_names
        apply_activation_checkpointing(model, check_fn=check)

def train(args):
    rank = int(os.environ["RANK"])
    local_rank = int(os.environ["LOCAL_RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
    if args.no_ddp and world_size > 1:
        raise ValueError("Distributed training is enabled but WORLD_SIZE > 1. Use --no_ddp to disable DDP.")
    dist.init_process_group(backend='nccl')
    device = local_rank # torch.device(f'cuda:{local_rank}')
    if rank == 0: print("Starting Aurora training script with arguments:", args)
    
    # Dummy dataset
    N = 5  # Number of samples
    M = 69 # 4 + 5 * 13
    H, W = 720, 1440
    lat=torch.linspace(90, -90, H)
    lon=torch.linspace(0, 360, W+1)[:-1]
    random_input = torch.randn(N, 2, M, H, W, dtype=torch.bfloat16)
    random_target = torch.randn(N, 1, M, H, W, dtype=torch.bfloat16)
    dataset = TensorDataset(random_input, random_target)
    sampler = DistributedSampler(dataset, shuffle=True)
    dataloader = DataLoader(dataset, sampler=sampler, batch_size=1, num_workers=1)
    if rank == 0: print("Dummy dataset created with random data.")
    
    # model
    if args.small: model = AuroraSmallPretrained(bf16_mode=args.bf16, use_lora=False)
    else: model = Aurora(bf16_mode=args.bf16, use_lora=False)
    model.load_checkpoint(strict=False)
    custom_activation_checkpointing(model, args.checkpointing_module_names)
    model = model.to(device)
    model = model.train()
    if not args.no_ddp:
        model = DDP(model, device_ids=[local_rank])
    if rank == 0: print("Model loaded and wrapped in DDP.")
    
    # loss and optimizer
    criterion = nn.L1Loss().to(device)
    optimizer = torch.optim.AdamW(model.parameters(), lr=5e-5, weight_decay=5e-6)
    if rank == 0: print("Loss and optimizer configured.")
    
    # autocast context
    if args.autocast: autocast_context = torch.autocast(device_type='cuda', dtype=torch.bfloat16)
    else: autocast_context = contextlib.nullcontext()
    
    # training loop
    global_step = 0
    for epoch in range(args.num_epochs):
        pbar = tqdm(dataloader, desc=f"Epoch {epoch+1}/{args.num_epochs}", disable=(local_rank != 0))
        for _, (inputs, targets) in enumerate(dataloader):
            # to batch and move to device
            inputs, targets = map(lambda x: make_batch(x, lat, lon)._fmap(lambda x: x.to(device, non_blocking=True)), (inputs, targets))
            optimizer.zero_grad()
            # fwd
            with autocast_context:
                outputs = model(inputs)
                loss = compute_loss(criterion, outputs, targets)
            # bwd 
            loss.backward()
            optimizer.step()
            # print
            pbar.update(1)
            pbar.set_postfix(loss=loss.detach().item(), step=global_step)
            global_step += 1
                
def main(args):
    try: train(args)
    except Exception as e: dist.destroy_process_group() ; raise e
    
if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--num_epochs", type=int, default=10, help="Number of epochs to train for")
    parser.add_argument("--small", action="store_true", help="Use the small version of the Aurora model")
    parser.add_argument("--bf16", action="store_true", help="Use `bf16_mode`")
    parser.add_argument("--autocast", action="store_true", help="Use autocast context with bf16")
    parser.add_argument("--checkpointing_module_names", type=str, nargs='+', 
                        default=["Perceiver3DEncoder","Swin3DTransformerBackbone","Basic3DEncoderLayer","Basic3DDecoderLayer","Perceiver3DDecoder","LinearPatchReconstruction"])
    parser.add_argument("--no_ddp", action="store_true", help="Disable DDP")
    args = parser.parse_args()
    main(args)
```

## Launch command

```bash
 srun torchrun --standalone --nnodes=1 --nproc_per_node=2 \
            train.py --num_epochs=500 --bf16 \
                --checkpointing_module_names Perceiver3DEncoder Swin3DTransformerBackbone Basic3DEncoderLayer Basic3DDecoderLayer Perceiver3DDecoder LinearPatchReconstruction
```

Does that look reasonable to you? I would appreciate any recommendations you might have :) Thanks!

**Comment by wesselb** [comment]
Hey @eliotwalt! I'm sorry you're running into these issues. :( Fitting the model onto the GPU was one of the main challenges in the project.

For the original training, we didn't use `bf16_mode` (which is experimental and I had hoped would solve these issues, but only seems giving memory access errors)  but (1) `autocast` to `torch.bfloat16`, (2) `FSDP` to free up extra memory on the GPUs, and (3) checkpointing only over `Basic3DEncoderLayer Basic3DDecoderLayer`. That should be the most robust setup, but it's unclear whether that will actually fit, as the whole things is super sensitive to the details (including PyTorch version) and the open-source version has diverged from our internal implementation. The other thing that could help is to use a better attention kernel (see [here](https://github.com/microsoft/aurora/issues/119#issuecomment-3048684287)). I'm still hoping that there should be some way of making things fit on one GPU without resorting to pure bf16.

I'm very sorry to say that I have no better suggestions. One possibility is to just run the whole architecture in bf16, which should work but which might come with significant stability issues.

I would love to help out with this, but unfortunately I currently don't have the time or resources to work on this.

@crisbodnar, do you have any other suggestions that could help?

**Comment by eliotwalt** [comment]
Hey @wesselb, thanks for your detailed reply. We were finally able to get stable training and inference using an ngc container (`nvcr.io/nvidia/pytorch:25.06-py3`). It seems that in the end, it was indeed an environment issue.

**Comment by eliotwalt** [comment]
Sure @wesselb, happy to share the script to build the container. It is specific to our HPC but people can probably adapt it. We use `apptainer` but I think other tools (e.g. `docker`) are supported (see [official Nvidia ngc docs](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch))
```bash
CONTAINER_URL=nvcr.io/nvidia/pytorch:25.06-py3
CONTAINER_NAME=torch-2.7-aurora-25-06.sif

# Optionally define output dir
OUTPUT_DIR=

if [ -z "$OUTPUT_DIR" ]; then
  echo "after if"
  # default output dir is filepath of this file
  OUTPUT_DIR=$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")
fi
echo "OUTPUT_DIR: $OUTPUT_DIR"

CONTAINER_OUTPUT_PATH=$OUTPUT_DIR/$CONTAINER_NAME

export APPTAINER_TMPDIR=/dev/shm/$USER
export APPTAINER_CACHEDIR=/scratch-shared/$USER/apptainer

mkdir -p $OUTPUT_DIR $APPTAINER_TMPDIR $APPTAINER_CACHEDIR $CONTAINER_OUTPUT_PATH

# Define apptainer definition file inline for more flexibility
TMP_CONTAINER_FILENAME=/dev/shm/$USER/torch.def

cat <<EOF > $TMP_CONTAINER_FILENAME
Bootstrap: docker
From: $CONTAINER_URL

%post
     export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda-12.9/compat
     export PIP_NO_CACHE_DIR=1
     pip install microsoft-aurora==1.7.0 timm

%environment
    export LD_PRELOAD=
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda-12.9/compat
    
EOF

echo "Downloading $CONTAINER_URL to $CONTAINER_OUTPUT_PATH"
apptainer build --force $CONTAINER_OUTPUT_PATH $TMP_CONTAINER_FILENAME

rm $TMP_CONTAINER_FILENAME
echo "Done building! Check out $CONTAINER_OUTPUT_PATH"
```

Then, we only need to run any script from inside the container. 
```bash
# start and enter container
apptainer shell --nv $CONTAINER_OUTPUT_PATH

# run script
python some_aurora_script.py 
```

I haven't tested this on another setup but I hope that it can be helpful.

**Comment by wesselb** [comment]
@eliotwalt, thanks for sharing this! That's incredibly helpful. :)

Which settings you eventually use for precision, checkpointing, and distributed computation? Plain DDP with BF16 AMP and checkpointing only over `Basic3DEncoderLayer Basic3DDecoderLayer`?

**Comment by eliotwalt** [comment]
@wesselb, at the moment we only ensured that the computations would run without errors. We used plain DDP (similar to the script above) to distribute the computation, in conjunction with torchrun and slurm. We used the default activation checkpointing and wrapped the forward path and loss in autocast with bf16.

**Comment by wesselb** [comment]
@eliotwalt Thanks, that's great to hear! By ensuring that the computations run without errors, do you mean that you also did a few gradient updates like the above script does?

If this setup turns out to work well for you, perhaps we can add it to the documentation!

**Comment by eliotwalt** [comment]
@wesselb, yes that's what I mean. I updated the default branch of the test repository that we have setup. You can find the training and launching scripts [here](https://github.com/eliotwalt/aurora-testbed/blob/container/train.py) and [here](https://github.com/eliotwalt/aurora-testbed/blob/container/train.job), respectively. At this point, I should give the credits to @robogast who came up with that brilliant solution! :) 

I think it would be good to have a way to replicate a stable environment in the documentation and indeed this container could be one way. Do you have a machine to test it on? Happy to assist as well as I can.

**Comment by wesselb** [comment]
Hey @eliotwalt! Apologies for the slow reply here.

> Concretely, I think it would be feasible to adapt the repo mentioned above and provide simple training and inference python scripts along with a container building script. You could then add that to an example/ folder or something like that. Then, the community can test the setup in different environments. How does that sound?

I think this would be fantastic. :)

I hope to soon have some spare time to work on this.
